### ISSUE TITLE
Pip resolves pandas install differently than poetry

### ISSUE THREAD

**Issue by fredrikaverpil** [python-poetry/poetry#4856]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [ x I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [ ] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: macOS 12.0.1 arm64
- **Poetry version**: 1.1.12
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: [](https://gist.github.com/fredrikaverpil/8052360bb121b8663ba17a1dee231211)

## Issue

I have a weird one. Given this `pyproject.toml`, a `poetry install`  (poetry 1.1.12, python 3.10.0 x86_64) will always install sub-dependency `numpy` of version `1.21.1`.
But version `1.21.4` of `numpy` is available. What gives?

```toml
[tool.poetry]
name = "pandastest"
version = "0.1.0"
description = ""
authors = []

[tool.poetry.dependencies]
python = "^3.10"
pandas = "^1.3.4"

[tool.poetry.dev-dependencies]

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

A regular old `pip install pandas` give me `numpy` `1.21.4`. Why does poetry resolve differently?

I tried adding it manually:

<details>
  <summary>Click to expand!</summary>
  ```bash
$ poetry add numpy@latest
Using version ^1.21.4 for numpy

Updating dependencies
Resolving dependencies... (0.0s)

  SolverProblemError

  The current project's Python requirement (>=3.10,<4.0) is not compatible with some of the required packages Python requirement:
    - numpy requires Python >=3.7,<3.11, so it will not be satisfied for Python >=3.11,<4.0
  
  Because no versions of numpy match >1.21.4,<2.0.0
   and numpy (1.21.4) requires Python >=3.7,<3.11, numpy is forbidden.
  So, because pandastest depends on numpy (^1.21.4), version solving failed.

  at ~/.local/pipx/venvs/poetry/lib/python3.10/site-packages/poetry/puzzle/solver.py:241 in _solve
      237│             packages = result.packages
      238│         except OverrideNeeded as e:
      239│             return self.solve_in_compatibility_mode(e.overrides, use_latest=use_latest)
      240│         except SolveFailure as e:
    → 241│             raise SolverProblemError(e)
      242│ 
      243│         results = dict(
      244│             depth_first_search(
      245│                 PackageNode(self._package, packages), aggregate_package_nodes

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties
    
    For numpy, a possible solution would be to set the `python` property to ">=3.10,<3.11"

    https://python-poetry.org/docs/dependency-specification/#python-restricted-dependencies,
    https://python-poetry.org/docs/dependency-specification/#using-environment-markers
  ```
</details>

Please note this part:

> - numpy requires Python >=3.7,<3.11, **_so it will not be satisfied for Python >=3.11,<4.0_**

**Comment by fredrikaverpil** [comment]
The solution was to edit this:

```toml
[tool.poetry.dependencies]
python = "^3.10"
```

into this:

```toml
[tool.poetry.dependencies]
python = ">=3.10, <3.11"
```

...and I really don't get why.

**Comment by finswimmer** [comment]
`poetry` and `pip` have different goals. While `pip` only cares about your current environment, `poetry` cares about your project definition and not about your current environment. 

Having a project that aims to be compatible with python `^3.10` means every python `>=3.10,<4.0`. This can only be true if all dependencies are available for this versions too. The latest numpy version is only available in the range >=3.7,<3.11. So it's not available for all the python versions you project aims to be compatible to and poetry is telling this to you.

Because pip doesn't care about any project definition and instead only on the current environment it will install the latest numpy for you.
